File: java/src/main/java/com/oltruong/Problem15.java
package com.oltruong;

import java.util.HashMap;
import java.util.Map;

class Problem15 {

    static long getResult() throws Exception {
        return findLatticePaths(20, 20, new HashMap<>());
    }

    private static long findLatticePaths(long x, long y, Map<String, Long> knownPaths) {
        final String key = x + "-" + y;
        if (x == 0 || y == 0) {
            return 1;
        } else if (knownPaths.containsKey(key)) {
            return knownPaths.get(key);
        } else {
            long result = findLatticePaths(x - 1, y, knownPaths) + findLatticePaths(x, y - 1, knownPaths);
            knownPaths.put(key, result);
            return result;
        }
    }
}

File: python/main/problem15.py
def get_problem15():
    return find_all_lattice_paths(20)


def find_all_lattice_paths(size):
    known_paths = dict()
    return find_lattice_paths(size, size, known_paths)


def find_lattice_paths(x, y, known_paths):
    key = f"{x}-{y}"
    if key in known_paths:
        return known_paths[key]
    elif y == 0 or x == 0:
        return 1
    else:
        result = find_lattice_paths(x, y - 1, known_paths) + find_lattice_paths(x - 1, y, known_paths)
        known_paths[key] = result
        return result
